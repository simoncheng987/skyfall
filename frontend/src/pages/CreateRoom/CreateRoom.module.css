.title {
  margin-bottom: 15px;
}

.textField {
  width: 100%;
  margin-bottom: 20px;
}

.button {
  width: 100%;
  height: 70px;
}

.toast {
  position: absolute;
  left: 50%;
  bottom: 5%;
  transform: translateX(-50%);
}
