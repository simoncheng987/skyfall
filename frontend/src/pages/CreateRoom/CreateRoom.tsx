import React, { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { io } from 'socket.io-client';
import PageScaffold from '../PageScaffold';
import Title from '../../components/Title';
import TextField from '../../components/TextField';
import Button from '../../components/Button';
import styles from './CreateRoom.module.css';
import ErrorToast from '../../components/ErrorToast';
import { useClient } from '../../context/ClientProvider';
import BackButton from '../../components/BackButton';
import SpinBox from '../../components/SpinBox';
import Dropdown from '../../components/Dropdown';

/**
 * This is a page for creating a room for playing the game.
 * Players can pick a list of word and configure the starting lives in this page.
 */
export default function CreateRoom() {
  const ERROR_INVALID_NAME = 'Please input your name.';
  const ERROR_MISSING_ROOM = 'Ohho! Cannot create room.';
  const WORD_LIST_NAME = 'default';

  const [playerName, setPlayerName] = useState('');
  const [error, setError] = useState(false);
  const [errorMessage, setErrorMessage] = useState('');
  const [roomId, setRoomId] = useState('');
  const [lives, setInitialLives] = useState(3);
  const [listNames, setListNames] = useState<string[]>([WORD_LIST_NAME]);
  const [pickedList, setPickedList] = useState<string>(WORD_LIST_NAME);

  const navigate = useNavigate();

  const { client, setClient, setName, setHost } = useClient();

  useEffect(() => {
    // Fetching the list of word for player to pick
    fetch('/wordList')
      .then((res) => res.json())
      .then((data) =>
        data.map((wordList: { listName: string; wordList: string[] }) => wordList.listName),
      )
      .then((wordlistNames) => setListNames(wordlistNames))
      .catch(() => {});
  }, []);

  useEffect(() => {
    const timeout = setInterval(() => {
      setError(false);
    }, 2000);

    return () => {
      clearTimeout(timeout);
    };
  }, [error]);

  useEffect(() => {
    // Setting up the client when it is available.
    if (client) {
      client.once('room:join-success', () => {
        setName(playerName);
        setHost();
        navigate('/lobby', { state: { lives, pickedList, roomId } });
      });

      client.once('room:join-fail', () => {
        setErrorMessage(ERROR_MISSING_ROOM);
        setError(true);
      });

      client.emit('room:join', roomId, playerName);

      return () => {
        client.off('room:join-success');
        client.off('room:join-fail');
      };
    }
    return () => {};
  }, [client]);

  function createRoom() {
    if (playerName.trim().length === 0) {
      setErrorMessage(ERROR_INVALID_NAME);
      setError(true);
    } else {
      fetch('/game', {
        method: 'POST',
      })
        .then((res) => res.text())
        .then((room) => {
          setRoomId(room);
          setClient(io());
        })
        .catch(() => {
          setErrorMessage(ERROR_MISSING_ROOM);
          setError(true);
        });
    }
  }

  return (
    <PageScaffold>
      <BackButton />

      <Title className={styles.title} fontSize="110px" text="Create Room" colorScheme="pink" />

      <h1 className={styles.label}>Word List</h1>
      <Dropdown options={listNames} onChange={(e) => setPickedList(e.target.value)} />

      <h1 className={styles.label}>Player Lives</h1>
      <SpinBox
        className={styles.spinbox}
        min={1}
        max={6}
        defaultValue={3}
        onChange={(e) => setInitialLives(parseInt(e.target.value, 10))}
      />

      <TextField
        className={styles.textField}
        placeholder="Your Name"
        onChange={(e) => setPlayerName(e.target.value)}
        ariaLabel="name-field"
      />

      <Button
        className={styles.button}
        text="Enter"
        onClick={() => createRoom()}
        ariaLabel="enter-button"
      />
      {error && <ErrorToast className={styles.toast} text={errorMessage} />}
    </PageScaffold>
  );
}
