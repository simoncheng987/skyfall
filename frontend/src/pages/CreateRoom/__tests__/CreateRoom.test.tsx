import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import ClientContextProvider from '../../../context/ClientProvider';
import CreateRoom from '../CreateRoom';

describe('CreateRoom', () => {
  it('matches snapshot', () => {
    const tree = renderer
      .create(
        <ClientContextProvider>
          <MemoryRouter initialEntries={[{ pathname: '/create', key: 'testKey' }]}>
            <CreateRoom />
          </MemoryRouter>
        </ClientContextProvider>,
      )
      .toTree();
    expect(tree).toMatchSnapshot();
  });
});
