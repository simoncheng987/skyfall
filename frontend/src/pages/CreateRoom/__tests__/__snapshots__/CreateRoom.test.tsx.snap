// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateRoom matches snapshot 1`] = `
Object {
  "instance": null,
  "nodeType": "component",
  "props": Object {
    "children": <MemoryRouter
      initialEntries={
        Array [
          Object {
            "key": "testKey",
            "pathname": "/create",
          },
        ]
      }
    >
      <CreateRoom />
    </MemoryRouter>,
  },
  "rendered": Object {
    "instance": null,
    "nodeType": "component",
    "props": Object {
      "children": <CreateRoom />,
      "initialEntries": Array [
        Object {
          "key": "testKey",
          "pathname": "/create",
        },
      ],
    },
    "rendered": Object {
      "instance": null,
      "nodeType": "component",
      "props": Object {
        "basename": undefined,
        "children": <CreateRoom />,
        "location": Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/create",
          "search": "",
          "state": null,
        },
        "navigationType": "POP",
        "navigator": Object {
          "action": "POP",
          "back": [Function],
          "block": [Function],
          "createHref": [Function],
          "forward": [Function],
          "go": [Function],
          "index": 0,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/create",
            "search": "",
            "state": null,
          },
          "push": [Function],
          "replace": [Function],
        },
      },
      "rendered": Object {
        "instance": null,
        "nodeType": "component",
        "props": Object {},
        "rendered": Object {
          "instance": null,
          "nodeType": "component",
          "props": Object {
            "children": Array [
              <Title
                className="title"
                colorScheme="pink"
                fontSize="110px"
                text="Create Room"
              />,
              <TextField
                className="textField"
                onChange={[Function]}
                placeholder="Your Name"
              />,
              <Button
                buttonStyle="primary"
                className="button"
                onClick={[Function]}
                text="Enter"
              />,
              false,
            ],
          },
          "rendered": Object {
            "instance": null,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="children"
                >
                  <div
                    className="error"
                  >
                    <ForwardRef(ExclamationCircleIcon)
                      width={35}
                    />
                    <p>
                      Oh no! Ensure that your monitor size is at least 1280 x 720, and display ratio is greater than 16:10
                    </p>
                  </div>
                </div>,
                <img
                  alt="grass background"
                  className="grassImage"
                  src="/assets/grass.png"
                />,
              ],
              "className": "container",
            },
            "rendered": Array [
              Object {
                "instance": null,
                "nodeType": "host",
                "props": Object {
                  "children": <div
                    className="error"
                  >
                    <ForwardRef(ExclamationCircleIcon)
                      width={35}
                    />
                    <p>
                      Oh no! Ensure that your monitor size is at least 1280 x 720, and display ratio is greater than 16:10
                    </p>
                  </div>,
                  "className": "children",
                },
                "rendered": Array [
                  Object {
                    "instance": null,
                    "nodeType": "host",
                    "props": Object {
                      "children": Array [
                        <ForwardRef(ExclamationCircleIcon)
                          width={35}
                        />,
                        <p>
                          Oh no! Ensure that your monitor size is at least 1280 x 720, and display ratio is greater than 16:10
                        </p>,
                      ],
                      "className": "error",
                    },
                    "rendered": Array [
                      Object {
                        "instance": null,
                        "nodeType": "host",
                        "props": Object {
                          "aria-hidden": "true",
                          "children": <path
                            clipRule="evenodd"
                            d="M18 10a8 8 0 11-16 0 8 8 0 0116 0zm-7 4a1 1 0 11-2 0 1 1 0 012 0zm-1-9a1 1 0 00-1 1v4a1 1 0 102 0V6a1 1 0 00-1-1z"
                            fillRule="evenodd"
                          />,
                          "fill": "currentColor",
                          "viewBox": "0 0 20 20",
                          "width": 35,
                          "xmlns": "http://www.w3.org/2000/svg",
                        },
                        "rendered": Array [
                          Object {
                            "instance": null,
                            "nodeType": "host",
                            "props": Object {
                              "clipRule": "evenodd",
                              "d": "M18 10a8 8 0 11-16 0 8 8 0 0116 0zm-7 4a1 1 0 11-2 0 1 1 0 012 0zm-1-9a1 1 0 00-1 1v4a1 1 0 102 0V6a1 1 0 00-1-1z",
                              "fillRule": "evenodd",
                            },
                            "rendered": Array [],
                            "type": "path",
                          },
                        ],
                        "type": "svg",
                      },
                      Object {
                        "instance": null,
                        "nodeType": "host",
                        "props": Object {
                          "children": "Oh no! Ensure that your monitor size is at least 1280 x 720, and display ratio is greater than 16:10",
                        },
                        "rendered": Array [
                          "Oh no! Ensure that your monitor size is at least 1280 x 720, and display ratio is greater than 16:10",
                        ],
                        "type": "p",
                      },
                    ],
                    "type": "div",
                  },
                ],
                "type": "div",
              },
              Object {
                "instance": null,
                "nodeType": "host",
                "props": Object {
                  "alt": "grass background",
                  "className": "grassImage",
                  "src": "/assets/grass.png",
                },
                "rendered": Array [],
                "type": "img",
              },
            ],
            "type": "div",
          },
          "type": [Function],
        },
        "type": [Function],
      },
      "type": [Function],
    },
    "type": [Function],
  },
  "type": [Function],
}
`;
